File: notebooks/client_utils.py
# Copyright © Amazon.com and Affiliates: This deliverable is considered Developed Content as defined in the AWS Service
# Terms and the SOW between the parties dated 2025.

"""Client utils for the api_gateway_demo notebook."""
import json
import logging

import requests


def submit_job(
    api_url: str,
    job_name: str,
    job_type: str,
    works: list,
    api_key: str,
) -> dict:
    """Submit job."""
    # Construct the full URL
    api_url = api_url.rstrip("/")
    endpoint = f"{api_url}/create_job"

    # Headers
    headers = {
        "Content-Type": "application/json",
        "x-api-key": api_key,
    }
    request_body = {"job_name": job_name, "job_type": job_type, "works": works}

    # Make the POST request
    response = requests.post(endpoint, data=json.dumps(request_body), headers=headers)

    # Check if the request was successful
    if response.status_code == 200:
        # Parse the JSON response
        return response.json()
    else:
        logging.error(f"Error: API request failed with status code {response.status_code}")
        logging.error(f"Response: {response.text}")
        response.raise_for_status()


def get_job_progress(api_url: str, job_name: str, api_key: str) -> dict:
    """Query the job_progress endpoint with the given job_name.

    Args:
    api_url (str): The base URL of your API Gateway
    job_name (str): The name of the job to query

    Returns:
    dict: The JSON response from the API, or None if an error occurred
    """
    api_url = api_url.rstrip("/")
    # Construct the full URL
    endpoint = f"{api_url}/job_progress"

    # Set up the query parameters
    params = {"job_name": job_name}

    # Headers
    headers = {"x-api-key": api_key}

    # Make the GET request
    response = requests.get(endpoint, params=params, headers=headers)

    # Check if the request was successful
    if response.status_code == 200:
        # Parse the JSON response
        return response.json()
    # Check if the item was not found
    if response.status_code == 404:
        # Parse the JSON response
        logging.error(f"Error: API request failed with status code {response.status_code}")
        return response.json()
    else:
        logging.error(f"Error: API request failed with status code {response.status_code}")
        logging.error(f"Response: {response.text}")
        response.raise_for_status()


def get_overall_progress(api_url: str, api_key: str) -> dict:
    """Query the overall_progress endpoint

    Args:
    api_url (str): The base URL of your API Gateway

    Returns:
    dict: The JSON response from the API, or None if an error occurred
    """
    api_url = api_url.rstrip("/")
    # Construct the full URL
    endpoint = f"{api_url}/overall_progress"

    # Headers
    headers = {"x-api-key": api_key}

    # Make the GET request
    response = requests.get(endpoint, headers=headers)

    # Check if the request was successful
    if response.status_code == 200:
        # Parse the JSON response
        return response.json()
    else:
        logging.error(f"Error: API request failed with status code {response.status_code}")
        logging.error(f"Response: {response.text}")
        response.raise_for_status()


def get_job_results(api_url: str, job_name: str, work_id: str, api_key: str) -> dict:
    """
    Query the results endpoint with the given job_name and work_id.

    Args:
    api_url (str): The base URL of your API Gateway.
    job_name (str): The name of the job to query.
    work_id (str): The ID of the work within the job.

    Returns:
    dict: The JSON response from the API, or None if an error occurred
    """
    api_url = api_url.rstrip("/")
    # Construct the full URL
    endpoint = f"{api_url}/results"

    # Set up the query parameters
    params = {"job_name": job_name, "work_id": work_id}

    # Headers
    headers = {"x-api-key": api_key}

    # Make the GET request
    response = requests.get(endpoint, params=params, headers=headers)

    # Check if the request was successful
    if response.status_code == 200:
        # Parse the JSON response
        return response.json()
    else:
        logging.error(f"Error: API request failed with status code {response.status_code}")
        logging.error(f"Response: {response.text}")
        response.raise_for_status()


def update_job_results(api_url: str, job_name: str, work_id: str, api_key: str) -> dict:
    # Construct the full URL
    api_url = api_url.rstrip("/")
    endpoint = f"{api_url}/results"

    # Headers
    headers = {
        "Content-Type": "application/json",
        "x-api-key": api_key,
    }

    updated_fields = {"work_status": "REVIEWED"}
    request_body = {"job_name": job_name, "work_id": work_id, "updated_fields": updated_fields}

    # Make the POST request
    response = requests.post(endpoint, data=json.dumps(request_body), headers=headers)

    # Check if the request was successful
    if response.status_code == 200:
        # Parse the JSON response
        return response.json()
    else:
        logging.error(f"Error: API request failed with status code {response.status_code}")
        logging.error(f"Response: {response.text}")
        response.raise_for_status()

File: lib/ruby/src/image_captioning_assistant/client/client_utils.rb
# Copyright © Amazon.com and Affiliates: This deliverable is considered Developed Content as defined in the AWS Service
# Terms and the SOW between the parties dated 2025.

require 'json'
require 'logger'
require 'net/http'
require 'uri'

# Client utils for the api_gateway_demo notebook.
class ApiClient
  def initialize
    @logger = Logger.new(STDOUT)
  end

  def submit_job(api_url, job_name, job_type, works, api_key)
    # Construct the full URL
    api_url = api_url.sub(%r{/+$}, '')
    endpoint = "#{api_url}/create_job"
    uri = URI(endpoint)

    # Headers and body
    headers = {
      'Content-Type' => 'application/json',
      'x-api-key' => api_key
    }
    request_body = {
      job_name: job_name,
      job_type: job_type,
      works: works
    }.to_json

    # Make the POST request
    response = make_request(Net::HTTP::Post, uri, headers, request_body)
    JSON.parse(response.body)
  end

  def get_job_progress(api_url, job_name, api_key)
    # Construct the full URL
    api_url = api_url.sub(%r{/+$}, '')
    endpoint = "#{api_url}/job_progress"
    uri = URI(endpoint)
    uri.query = URI.encode_www_form(job_name: job_name)

    # Headers
    headers = { 'x-api-key' => api_key }

    # Make the GET request
    response = make_request(Net::HTTP::Get, uri, headers)

    if response.code == '404'
      @logger.error("Error: API request failed with status code #{response.code}")
      return JSON.parse(response.body)
    end

    JSON.parse(response.body)
  end

  def get_overall_progress(api_url, api_key)
    # Construct the full URL
    api_url = api_url.sub(%r{/+$}, '')
    endpoint = "#{api_url}/overall_progress"
    uri = URI(endpoint)

    # Headers
    headers = { 'x-api-key' => api_key }

    # Make the GET request
    response = make_request(Net::HTTP::Get, uri, headers)
    JSON.parse(response.body)
  end

  def get_job_results(api_url, job_name, work_id, api_key)
    # Construct the full URL
    api_url = api_url.sub(%r{/+$}, '')
    endpoint = "#{api_url}/results"
    uri = URI(endpoint)
    uri.query = URI.encode_www_form(job_name: job_name, work_id: work_id)

    # Headers
    headers = { 'x-api-key' => api_key }

    # Make the GET request
    response = make_request(Net::HTTP::Get, uri, headers)
    JSON.parse(response.body)
  end

  def update_job_results(api_url, job_name, work_id, api_key, updated_fields)
    # Construct the full URL
    api_url = api_url.sub(%r{/+$}, '')
    endpoint = "#{api_url}/results"
    uri = URI(endpoint)

    # Headers and body
    headers = {
      'Content-Type' => 'application/json',
      'x-api-key' => api_key
    }
    request_body = {
      job_name: job_name,
      work_id: work_id,
      updated_fields: updated_fields
    }.to_json

    # Make the PUT request
    response = make_request(Net::HTTP::Put, uri, headers, request_body)
    JSON.parse(response.body)
  end

  private

  def make_request(request_class, uri, headers, body = nil)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = (uri.scheme == 'https')

    request = request_class.new(uri)
    headers.each { |key, value| request[key] = value }
    request.body = body if body

    response = http.request(request)

    return response if response.is_a?(Net::HTTPSuccess)

    @logger.error("Error: API request failed with status code #{response.code}")
    @logger.error("Response: #{response.body}")

    return response if response.code == '404' && request_class == Net::HTTP::Get

    raise "API request failed: #{response.code} #{response.message}"
  end
end
